src/Simulation/Simulators/OpenSystemsSimulator/DataModel/Instrument.cs

```diff
@@ -3,90 +3,95 @@
-                _ => throw new JsonException($"Enum variant {variant} not yet supported.")
-            };
-
-            reader.Require(JsonTokenType.EndObject);
-
-            return result;
+            throw reader.CurrentState.Exception("Instrument variant name was null.");
         }
 
-        public override void Write(Utf8JsonWriter writer, Instrument value, JsonSerializerOptions options)
+        Instrument result = variant switch
         {
-
-            switch (value)
+            "Effects" => new EffectsInstrument
             {
-                case EffectsInstrument effectsInstrument:
-                    writer.WriteStartObject();
-                    writer.WritePropertyName("Effects");
-                    JsonSerializer.Serialize(writer, effectsInstrument.Effects);
-                    writer.WriteEndObject();
-                    break;
-
-                case ZMeasurementInstrument zInstrument:
-                    writer.WriteStartObject();
-                    writer.WritePropertyName("ZMeasurement");
-                    JsonSerializer.Serialize(writer, zInstrument);
-                    writer.WriteEndObject();
-                    break;
-
-                default:
-                    throw new JsonException($"Enum variant {value.GetType()} not yet supported.");
-            }
-
-        }
+                Effects = JsonSerializer
+                    .Deserialize<List<Process>>(ref reader)
+                    .ToImmutableList()
+            },
+            "ZMeasurement" =>
+                JsonSerializer.Deserialize<ZMeasurementInstrument>(ref reader)
+                is {} zMeas
+                ? zMeas
+                : throw reader.CurrentState.Exception("Failed to deserialize Z measurement instrument."),
+            _ => throw reader.CurrentState.Exception($"Enum variant {variant} not yet supported.")
+        };
+
+        reader.Require(JsonTokenType.EndObject);
+
+        return result;
     }
 
-    [JsonConverter(typeof(InstrumentConverter))]
-    public abstract class Instrument
+    public override void Write(Utf8JsonWriter writer, Instrument value, JsonSerializerOptions options)
     {
 
+        switch (value)
+        {
+            case EffectsInstrument effectsInstrument:
+                writer.WriteStartObject();
+                writer.WritePropertyName("Effects");
+                JsonSerializer.Serialize(writer, effectsInstrument.Effects);
+                writer.WriteEndObject();
+                break;
+
+            case ZMeasurementInstrument zInstrument:
+                writer.WriteStartObject();
+                writer.WritePropertyName("ZMeasurement");
+                JsonSerializer.Serialize(writer, zInstrument);
+                writer.WriteEndObject();
+                break;
+
+            default:
+                throw new JsonException($"Enum variant {value.GetType()} not yet supported.");
+        }
```

Comment:
(Minor optimization and readability) Consider extracting the parts common for `case` arms to outside of `switch`:
```suggestion
        writer.WriteStartObject();
        switch (value)
        {
            case EffectsInstrument effectsInstrument:
                writer.WritePropertyName("Effects");
                JsonSerializer.Serialize(writer, effectsInstrument.Effects);
                break;

            case ZMeasurementInstrument zInstrument:
                writer.WritePropertyName("ZMeasurement");
                JsonSerializer.Serialize(writer, zInstrument);
                break;

            default:
                throw new JsonException($"Enum variant {value.GetType()} not yet supported.");
        }
        writer.WriteEndObject();
```

Reply:
I'd prefer not to in this case for a couple reasons:

- In some cases, enum variants may be serialized as a single string rather than as an object (in particular, that's the case for variants of type unit).
- Doing so would leave the writer in an inconsistent state if an exception occurs (e.g.: the exception thrown by `default`), as an object would then be started but not ended.